File: 20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
  const leftSymbols = ["(", "[", "{"];
  const rightSymbols = [")", "]", "}"];

  const stack: string[] = [];

  for (let char of s) {
    if (leftSymbols.includes(char)) {
      stack.push(char);
    } else {
      if (
        leftSymbols.indexOf(stack[stack.length - 1]) ===
        rightSymbols.indexOf(char)
      ) {
        stack.pop();
      } else {
        return false;
      }
    }
  }

  return stack.length === 0;
}
// @lc code=end

function BetterIsValid(s: string): boolean {
  const bracketsMap = {
    ")": "(",
    "]": "[",
    "}": "{",
  };

  const stack: string[] = [];

  for (let i = 0; i < s.length; i++) {
    const curChar = s[i];
    if (["(", "[", "{"].includes(curChar)) {
      stack.push(curChar);
    } else {
      if (stack.pop() !== bracketsMap[curChar]) {
        return false;
      }
    }
  }

  return !stack.length;
}

File: 20.valid-parentheses.cpp
/*
 * @lc app=leetcode id=20 lang=cpp
 * @lcpr version=30204
 *
 * [20] Valid Parentheses
 */


// @lcpr-template-start
using namespace std;
#include <algorithm>
#include <array>
#include <bitset>
#include <climits>
#include <deque>
#include <functional>
#include <iostream>
#include <list>
#include <queue>
#include <stack>
#include <tuple>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
// @lcpr-template-end
// @lc code=start
class Solution {
public:
    bool isValid(string s) {
        if (s.size() == 0 || s.size() == 1) return false;
        stack<char> stk;

        for (auto c : s) {
            if (c == '(' || c == '[' || c == '{') {
                stk.push(c);
            } else {
                if (stk.empty()) {
                    return false;
                }

                if ((c == ')' && stk.top() == '(') 
                    || (c == ']' && stk.top() == '[') 
                    || (c == '}' && stk.top() == '{')
                ) {
                    stk.pop();
                } else {
                    return false;
                }
            }
        }

        return stk.empty();
    }
};
// @lc code=end



/*
// @lcpr case=start
// "()"\n
// @lcpr case=end

// @lcpr case=start
// "()[]{}"\n
// @lcpr case=end

// @lcpr case=start
// "(]"\n
// @lcpr case=end

// @lcpr case=start
// "([])"\n
// @lcpr case=end

 */

